include/inviwo/core/algorithm/histogram1d.h

```diff
@@ -0,0 +1,200 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <inviwo/core/common/inviwocoredefine.h>
+#include <inviwo/core/datastructures/histogram.h>
+#include <inviwo/core/util/glmutils.h>
+#include <inviwo/core/util/glmcomp.h>
+#include <inviwo/core/util/glmmatext.h>
+
+#include <glm/common.hpp>
+
+#include <ranges>
+#include <vector>
+#include <array>
+#include <span>
+#include <tuple>
+#include <cmath>
+
+namespace inviwo::util {
+
+IVW_CORE_API std::vector<double> calculatePercentiles(const std::vector<size_t>& hist, dvec2 range,
+                                                      size_t sum);
+IVW_CORE_API Statistics calculateHistogramStats(const std::vector<size_t>& hist);
+
+namespace detail {
+
+/**
+ * Determine optimal number of bins of a histogram while considering \p bins as an upper bound
+ * for the number of bins
+ * 1. if there are more than 100 samples per bin based on data range, use \p bins without
+ *    further modification
+ * 2. for integral types:
+ *    a. set bins to data range if the number of bins exceeds the data range
+ *    b. otherwise adjust the number of bins so that the bin size is a whole number,
+ *       i.e. \f$ n \in \mathbb{N} \f$
+ * 3. for floating point types:
+ *    a. if number of bins exceeds |data range|, use original \p bins count
+ *    b. if the data range is non-fractional, i.e. whole numbers
+ *       [min, max], adjust the number of bins so that the bin size is a whole number
+ * 4. otherwise, use original \p bins count
+ *
+ * @tparam T       type of the underlying data
+ * @param dataMap  used for calculating the number of bins
+ * @param bins     upper bound for the number of bins
+ * @return optimal number of bins and effective data range, which can be slightly larger than the
+ *                 original data range
+ */
+template <typename T>
+std::pair<size_t, double> optimalBinCount(DataMapper dataMap, size_t bins) {
+    auto intRangeBins = [&](size_t irange) -> std::pair<size_t, double> {
+        const size_t m = (irange + bins) / bins;
+        const auto numBins =
+            static_cast<size_t>(std::ceil(static_cast<double>(irange) / static_cast<double>(m)));
+        const auto effectiveRange = static_cast<double>((numBins - 1) * m);
+        return {numBins, effectiveRange};
+    };
+    auto isNonFractional = [](double v) { return std::abs(v - std::round(v)) < 1.0e-12; };
+
+    const auto range = (dataMap.dataRange.y - dataMap.dataRange.x);
+    if (range / static_cast<double>(bins) > 100.0) {  // case 1
+        // sufficient samples per bin
+        return {bins, range};
+    }
+    if constexpr (std::is_floating_point_v<util::value_type_t<T>>) {
+        if (range < static_cast<double>(bins)) {  // case 3.a
+            return {bins, range};
+        }
+        if (isNonFractional(dataMap.dataRange.x) && isNonFractional(dataMap.dataRange.y)) {
```

Comment:
What would happen if dataRange is like 1E-13 to 5E-13
Should probably also check that min - max >= 1?

Reply:
That is a good point.